### ISSUE TITLE
[Question]: Slowly receiveing client in long-lived streaming

### ISSUE THREAD

**Issue by Marwin34** [Tradias/asio-grpc#94]
Hi Tradias,

Is there a way to detect a slowly receiving client in a long-lived streaming RPC in asio-grpc (or grpc itself)? I am working on a server application that sends data to clients using streaming RPCS, and I would like to detect swlowly receiving clients and then drop the connection with them since it can lead to queuing up data on sockets, potentially resulting in slowing down the server or ooming it. Could such a mechanism be implemented with asio-grpc?

Thanks in advance.

**Comment by Tradias** [comment]
Asio-grpc just acts as a small wrapper around gRPC's API. In the case of a client-streaming RPC, for example, that would be [grpc::ServerAsyncWriter::Write](https://grpc.github.io/grpc/cpp/classgrpc_1_1internal_1_1_async_writer_interface.html#a03f8532dfbd6c82c7d1fed5bc6e79d79). If that call really completes slowly for slow clients (I haven't tested it, also note this [set_write_through](https://grpc.github.io/grpc/cpp/classgrpc_1_1_write_options.html#aaa5f7fcba7816c5d6b41b3e62b2b9b58) option), then you can use it to automatically drop the connection like so:

```c++
const auto slow_threshold = std::chrono::seconds(5);
agrpc::Alarm alarm{grpc_context};
auto [completion_order, alarm_ok, write_ok] =
    co_await asio::experimental::make_parallel_group(
        alarm.wait(std::chrono::system_clock::now() + slow_threshold, asio::deferred),
        rpc.write(response, asio::deferred))
        .async_wait(asio::experimental::wait_for_one{}, asio::use_awaitable);
if (!write_ok)
{
    // client dropped connection or was slow (alarm finished and caused cancellation of rpc with invokes grpc::ServerContext::TryCancel)
    co_return;
}
```

**Comment by nqf** [comment]
How to call `set_write_through` in asio-grpc ?

**Comment by Tradias** [comment]
@nqf `ServerRPC::write`, `ServerRPC::write_and_finish` and `ClientRPC::write` all take an optional second argument:  [grpc::WriteOptions](https://grpc.github.io/grpc/cpp/classgrpc_1_1_write_options.html). Which means you can do:

```c++
rpc.write(request, grpc::WriteOptions().set_write_through(), <completion token>);
```

**Comment by nqf** [comment]
> ```c++
> rpc.write(request, grpc::WriteOptions().set_write_through(), <completion token>);
> ```

Is it possible for `rpc.write` to throw an exception?  As far as I know, ASIO can throw exceptions or return error codes

**Comment by Tradias** [comment]
No, since gRPC never throws exceptions I am not inventing any either. The completion signature of [rpc.write](https://tradias.github.io/asio-grpc/classagrpc_1_1_server_r_p_c_3_01_request_server_streaming_00_01_traits_t_00_01_executor_01_4.html#aa8ce64f00653fffa32d19c5ec76cda21) is `void(bool)` which differs from the typical asio completion signatures `void(error_code, <something>)`. Only if the first argument is an `error_code` will asio automatically convert it to an exception.

From the asio-grpc documentation:

> The completion signature is void(bool). true means that the data/metadata/status/etc is going to go to the wire. If it is false, it is not going to the wire because the call is already dead (i.e., canceled, deadline expired, other side dropped the channel, etc).

**Comment by nqf** [comment]
> No, since gRPC never throws exceptions I am not inventing any either. The completion signature of [rpc.write](https://tradias.github.io/asio-grpc/classagrpc_1_1_server_r_p_c_3_01_request_server_streaming_00_01_traits_t_00_01_executor_01_4.html#aa8ce64f00653fffa32d19c5ec76cda21) is `void(bool)` which differs from the typical asio completion signatures `void(error_code, <something>)`. Only if the first argument is an `error_code` will asio automatically convert it to an exception.
> 
> From the asio-grpc documentation:
> 
> > The completion signature is void(bool). true means that the data/metadata/status/etc is going to go to the wire. If it is false, it is not going to the wire because the call is already dead (i.e., canceled, deadline expired, other side dropped the channel, etc).

thanks

**Comment by nqf** [comment]
> @nqf Does this answer the op's question? Aka. can this issue be closed now?

I have no more questions，I think you can close it
